Add support for deployment annotations

mark-hofmeijer opened the pull request:
## What was changed
This PR adds support to define deployment annotations for the `server`, `admintools` and `web` components. The setup is similar to the recent implemented resource labels.

NOTE: this PR partly replaces #537, however this PR needs revision because of the templatization in PR #539.

## Why?
Currently there is no option to define deployment annotations.

## Checklist
1. How was this tested:
`helm install -f values.yaml --set server.deploymentAnnotations.deployAnnotation1="serverDeploymentAnnotation" --set admintools.deploymentAnnotations.deployAnnotation1="adminDeploymentAnnotation" --set web.deploymentAnnotations.deployAnnotation1="webDeploymentAnnotation" debug . --dry-run --debug`

- In the current situation, this adds no deployment annotations on any of the components.
- In the new situation, a deployment annotation `deployAnnotation1` is added to all server services and admintools and web components.

`helm install -f values.yaml --set server.deploymentAnnotations.serverAnnotation1="serverDeploymentAnnotation" --set server.frontend.deploymentAnnotations.resourceAnnotation1="resourceDeploymentAnnotation" debug . --dry-run --debug`

- In the current situation, this adds no deployment annotations on the server `frontend` service.
- In the new situation, a deployment annotation `resourceAnnotation1` is added to the server `frontend` service. The deployment annotations on server level are ignored correctly.
- In the new situation, a deployment annotation `serverAnnotation1` is added to the other server services (`history`, `matching`, `worker`). 

2. Any docs updates needed?
No.

robholland commented on charts/temporal/templates/web-deployment.yaml:
```diff
@@ -3,6 +3,10 @@ apiVersion: apps/v1
 kind: Deployment
 metadata:
   name: {{ include "temporal.componentname" (list $ "web") }}
+  {{- if or $.Values.additionalAnnotations $.Values.admintools.deploymentAnnotations }}
```
```suggestion
  {{- if or $.Values.additionalAnnotations $.Values.web.deploymentAnnotations }}
```

robholland replied:
On seconds thoughts lets remove the gating.

mark-hofmeijer replied:
Removed the gating.

robholland commented on charts/temporal/templates/_helpers.tpl:
```diff
@@ -71,6 +71,41 @@ Define the AppVersion
+{{- end -}}
+
+{{/*
+Additonal user specified annotations for all resources
```
Couldn't this just be inlined above? It's never called on it's own.

robholland replied:
Just noticed the label version is split out. That should be inlined also, but don't worry about that for this PR.

mark-hofmeijer replied:
It was indeed intended to align with the approach on the labels. But agree that inline is better, because it's not called on it's own.

mark-hofmeijer commented:
Refactored the helper method to the new approach introduced in #584. Also added a test.

Noted that a new line is added in front of the annotations (both on deployment and pod level), but did not manage to remove it although the setup looks the same as for labels. Maybe you can take a quick look @robholland ?

robholland commented on charts/temporal/templates/admintools-deployment.yaml:
```diff
@@ -3,6 +3,10 @@ apiVersion: apps/v1
 kind: Deployment
 metadata:
   name: {{ include "temporal.componentname" (list $ "admintools") }}
+  {{- if or $.Values.additionalAnnotations $.Values.admintools.deploymentAnnotations }}
```
Let's remove the gating, we can just allow `annotations:` to be empty, it's valid syntax.

mark-hofmeijer replied:
Removed the gating.

robholland commented on charts/temporal/tests/deployment_test.go:
```diff
@@ -104,3 +104,34 @@ func TestTemplateServerDeploymentMerging(t *testing.T) {
 	require.Equal(t, "three", deployment.Spec.Template.ObjectMeta.Labels["one"])
 	require.Equal(t, "two", deployment.Spec.Template.ObjectMeta.Labels["two"])
 }
+
+func TestTemplateServerDeploymentAnnotations(t *testing.T) {
+	// t.Parallel()
+
+	helmChartPath, err := filepath.Abs("../")
+	releaseName := "temporal"
+	require.NoError(t, err)
+
+	namespaceName := "temporal-" + strings.ToLower(random.UniqueId())
+
+	var deployment appsv1.Deployment
+
+	options := &helm.Options{
+		SetValues: map[string]string{
+			"server.frontend.deploymentAnnotations.one":  "three",
+			"server.frontend.deploymentAnnotations.four": "four",
+			"server.deploymentAnnotations.one":           "one",
+			"server.deploymentAnnotations.two":           "two",
+		},
+		KubectlOptions:    k8s.NewKubectlOptions("", "", namespaceName),
+		BuildDependencies: true,
+	}
+
+	output := helm.RenderTemplate(t, options, helmChartPath, releaseName, []string{"templates/server-deployment.yaml"})
+
+	helm.UnmarshalK8SYaml(t, output, &deployment)
+
+	require.Equal(t, "three", deployment.ObjectMeta.Annotations["one"])
+	require.Equal(t, "two", deployment.ObjectMeta.Annotations["two"])
+	require.Equal(t, "four", deployment.ObjectMeta.Annotations["four"])
+}
```
👍

mark-hofmeijer replied:
Did add a value test on the additional annotations as well.

robholland commented:
Please merge master to get the CI fix.
